src/effects/VST/VSTEffect.cpp

```diff
@@ -1026,60 +1041,91 @@ auto VSTEffectMessage::Clone() const -> std::unique_ptr<Message>
 {
    auto result = std::make_unique<VSTEffectMessage>(*this);
    // Make sure of the chunk capacity
-   result->settings.mChunk.reserve(this->settings.mChunk.capacity());
+   result->mChunk.reserve(this->mChunk.capacity());
+
    return result;
 }
 
 void VSTEffectMessage::Assign(Message && src)
 {
-   auto& dstSettings = this->settings;
-   auto& srcSettings = static_cast<VSTEffectMessage&>(src).settings;
+   VSTEffectMessage& vstSrc = static_cast<VSTEffectMessage&>(src);
+
+   mChunk = vstSrc.mChunk;
+   vstSrc.mChunk.resize(0);     // capacity will be preserved though
+
+   assert(mParamsVec.size() == vstSrc.mParamsVec.size());
+
+   for (size_t i = 0; i < mParamsVec.size(); i++)
+   {
+      mParamsVec[i] = vstSrc.mParamsVec[i];
 
-   VSTEffectWrapper::MoveSettingsContents(
-      std::move(srcSettings), dstSettings, /*merge=*/false);
+      // consume the source value
+      vstSrc.mParamsVec[i] = std::nullopt;
+   }
 }
 
 void VSTEffectMessage::Merge(Message && src)
 {
-   auto& dstSettings = this->settings;
+   VSTEffectMessage& vstSrc = static_cast<VSTEffectMessage&>(src);
 
-   auto& srcSettings = static_cast<VSTEffectMessage&>(src).settings;
+   bool chunkWasAssigned = false;
 
+   if ( ! vstSrc.mChunk.empty() )
+   {
+      mChunk = vstSrc.mChunk;
+      chunkWasAssigned = true;
+   }
+
+   vstSrc.mChunk.resize(0);  // capacity will be preserved though
+
+   assert(mParamsVec.size() == vstSrc.mParamsVec.size());
+
+   for (size_t i = 0; i < mParamsVec.size(); i++)
+   {
+      if (chunkWasAssigned)
+      {
+         mParamsVec[i] = vstSrc.mParamsVec[i];
+      }
+      else
+      {
+         // if src val is nullopt, do not copy it to dest
+         if (vstSrc.mParamsVec[i] != std::nullopt)
+         {
+            mParamsVec[i] = vstSrc.mParamsVec[i];
+         }
+      }
+
+      // consume the source value
+      vstSrc.mParamsVec[i] = std::nullopt;
+   }
 
-   VSTEffectWrapper::MoveSettingsContents(
-      std::move(srcSettings), dstSettings, /*merge=*/true);
 }
 
 
 std::unique_ptr<EffectInstance::Message> VSTEffectInstance::MakeMessage() const
 {
    VSTEffectSettings settings;
    FetchSettings(settings, false);
-   return std::make_unique<VSTEffectMessage>(std::move(settings));
-}
 
-
-std::unique_ptr<EffectInstance::Message> VSTEffectInstance::MakeMessage(int id, double value) const
-{
-   VSTEffectSettings settings;
+   VSTEffectMessage::ParamVector paramVector;
+   paramVector.resize(settings.mParamsMap.size(), std::nullopt);
 
    ForEachParameter
    (
```

Comment:
`FetchSettings` calls `ForEachParameter`.  Then you call `ForEachParameter` again.  And that means `GetString(effGetParamName, i)` is called `2 * mAEffect->numParams` times.

But the names are not needed, and the frequent calls of `GetString` might be a reason the main thread consumes too much CPU during slider movements.

So I suggest:  simply iterate an integer up to `mAEffect->numParams` and do `callGetParameter` directly.  It might help performance.

Reply:
You need a chunk too to make the message, so extract that part of `FetchSettings` as a function you can re-use here, with `doFetch == false` so that it only really reserves sufficient chunk space.

The purpose of the no-argument `MakeSettings()` is just initialization of Message objects with vectors of sufficient size.

Reply:
However, this code does not get called in response to a slider being moved, but only about 2 times when the effect is instantiated in the track, and then when its gui is opened. 

Anyway, I agree it is not so clean like it is now. I will do this too.

Reply:
However (again): the original code (previous to these changes) did not get the chunk **and** did not get the single values - so I should do the same here (which is consistent: either I get both kinds of info, or I get none), and for this I just need to remove the call to ForEachParam, which also solves the original objection.

Reply:
But a purpose of the call is to reserve sufficient space, to avoid memory allocations and dellocations later.

So if the loop is entirely eliminated, you should still make a vector of sufficient `nullopt`s.  That just requires the number of parameters.